osf/models/analytics.py

```diff
@@ -63,6 +65,26 @@ class PageCounter(BaseModel):
     total = models.PositiveIntegerField(default=0)
     unique = models.PositiveIntegerField(default=0)
 
+    DOWNLOAD_ALL_VERSIONS_ID_PATTERN = r'^download:[^:]*:{1}[^:]*$'
+
+    @classmethod
+    def get_all_downloads_on_date(cls, date):
+        """
+        Queries the total number of downloads on a date
+        :param str date: must be formatted the same as a page counter key so 'yyyy/mm/dd'
+        :return: long sum:
+        """
+        formatted_date = date.strftime('%Y/%m/%d')
+        # Get all PageCounters with data for the date made for all versions downloads,
+        # regex insures one colon so all versions are queried.
+        page_counters = cls.objects.filter(date__has_key=formatted_date, _id__regex=cls.DOWNLOAD_ALL_VERSIONS_ID_PATTERN)
+
+        # Get the total download numbers from the nested dict on the PageCounter by annotating it as daily_total then
+        # aggregating the sum.
+        daily_total = page_counters.annotate(daily_total=RawSQL("((date->%s->>'total')::int)", (formatted_date,))).aggregate(sum=Sum('daily_total'))['sum']
```

Comment:
You should be able to use `KeyTextTransform` here instead of `RawSQL`. Unfortunately, it's undocumented, but here's a thing: https://til.sloria.com/#2018-08-20-use-keytransform-to-query-on-jsonfields-in-django

Reply:
`KeyTextTransform` is very useful, but I think the RawSQL is necessary to cast the total as an int so it can be summed. With the query: `daily_total = page_counters.annotate(totals=KeyTextTransform(formatted_date, 'date')).aggregate(sum=Sum(KeyTextTransform('total', 'totals')))['sum']`

I'm getting the following:

```bash
 ProgrammingError: operator does not exist: text ->> unknown
 LINE 1: SELECT SUM(("totals" ->> 'total')) FROM (SELECT "osf_pagecou...
                              ^
 HINT:  No operator matches the given name and argument type(s). You might need to add explicit type casts.
```

Reply:
There's also a `KeyTransform` which I believe allows you to cast to any type.

Reply:
KeyTransform doesn't seem to do the trick, I'm getting the same problem. Looking at [the source](https://docs.djangoproject.com/en/2.1/_modules/django/contrib/postgres/fields/jsonb/) I can't see a obvious way of doing this with KeyTransform.

Reply:
Looking at the source of `Func` (from which `Transform` and `KeyTransform` inherit), it looks like you can pass `output_field`:

```python
KeyTransform('foo', 'bar', output_field=models.IntegerField())
```

Reply:
I've tried, the above and a few other approaches to casting this and they still bump into the same problem.